src/opendr/perception/semantic_segmentation/bisenet/bisenet_learner.py

```diff
@@ -0,0 +1,309 @@
+                print('Total Loss : %f' % (loss_train_mean))
+            if self.checkpoint_after_iter != 0 and epoch % self.checkpoint_after_iter == 0:
+                self.save(os.path.join(self.temp_path, 'epoch', str(epoch)))
+            if self.val_after != 0 and epoch % self.val_after == 0 and val_dataset is not None:
+                if not silent and verbose:
+                    print('Validation...')
+                self.eval(val_dataset)
+
+    def eval(self, dataset, silent=False, verbose=True):
+
+        dataloader_test = DataLoader(dataset,
+                                     shuffle=False,
+                                     pin_memory=self.pin_memory,
+                                     drop_last=False,
+                                     num_workers=self.num_workers)
+
+        with torch.no_grad():
+            self.model.eval()
+            precision_record = []
+            tq = tqdm.tqdm(total=len(dataloader_test) * 1)
+            tq.set_description('test')
+            hist = np.zeros((self.num_classes, self.num_classes))
+            for i, (data, label) in enumerate(dataloader_test):
+                tq.update(1)
+                if self.device == 'cuda':
+                    data = data.cuda()
+                    label = label.cuda()
+                predict = self.model(data).squeeze().cpu()
+                predict = reverse_one_hot(predict)
+                predict = np.array(predict)
+                label = label.squeeze().cpu()
+                label = np.array(label)
+                precision = compute_global_accuracy(predict, label)
+                hist += fast_hist(label.flatten(), predict.flatten(), self.num_classes)
+                precision_record.append(precision)
+            precision = np.mean(precision_record)
+            miou_list = per_class_iu(hist)[:-1]
+            miou = np.mean(miou_list)
+            tq.close()
+            if not silent and verbose:
+                print('precision for test: %.3f' % precision)
+                print('mIoU for validation: %.3f' % miou)
+
+            return precision, miou
+
+    def infer(self, img, csvpath, spath):
+        """
+        This method is used to perform semantic segmentation on an image.
+        It returns a heatmap of the given image.
+
+        """
+
+        if not isinstance(img, Image):
+            img = Image(img)
+        img = img.numpy()
+        image = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
+        resize = iaa.Scale({'height': self.crop_height, 'width': self.crop_width})
+        resize_det = resize.to_deterministic()
+        image = resize_det.augment_image(image)
+        image = PILImage.fromarray(image).convert('RGB')
+        image = transforms.ToTensor()(image)
+        image = transforms.Normalize((0.485, 0.456, 0.406), (0.229, 0.224, 0.225))(image).unsqueeze(0)
+        # read csv label path
+        label_info = get_label_info(os.path.join(csvpath, 'class_dict.csv'))
```

Comment:
This should be read during the creation of the object in order to avoid having re-supply the same labels and re-read the file each time we perform inference.